performance testing packages

adamdbrw opened the pull request:
Package for performance #435 metaticket tasks.

Builds with the rosbag2 project, but you will need additional dependencies. Please see [readme](https://github.com/RobotecAI/rosbag2/tree/master/rosbag2_performance/writer_benchmarking) for instructions on how to use it.

This is the first package out of 2 (plus one supporting) benchmarking packages for rosbag2. This one focuses on replicating and benchmarking the part when the message already arrives to the recorder through the ROS2 transport (so, doesn't include transport layer performance).

Current functionality:
- Contains scalable and configurable producers of data
- The build is controlled by a dedicated flag (BUILD_BENCHMARKS)
- Counts received messages and percentage of all received.
- Includes a simple convenience script to run batches with different parametrization and produce output files (which can be easily grepped and processed).

What is not there yet (proposed enhancements):
- automatic processing of output files from the batch run to produce a summary. It could be easily realized through use of grep or through adding a joint, csv (or similar) format file as an output.
- integration with an automated build processes
- support for compression and conversion. Both are relatively easy to add. 
- writer_benchmark doesn't need to be a ROS node. This could be removed with logging and parameters replace by non-ROS solutions. However, it uses rcl and rmw calls to mimic what is happening in the real recorder code (which is arguably worth enough not to make it ROS-agnostic).

adamdbrw commented:
Example partial result for a case with 10Kb messages with 100 topics at 100Hz. Note that there are some issues still and the summary report doesn't generate properly yet.

```
bechmark:
  title: 10KB with 100 topics
cpu:
  mean: 17.611134601832273
  std: 14.737801039292737
disk:
  read:
    mean: 0.0003239325842696629
    std: 0.003038756996757516
  write:
    mean: 94.42021651685393
    std: 53.916070906797025
memory:
  max: 6849.572864
  mean: 6590.834894387596
  min: 6412.9024
  std: 132.1152373612786
messages:
  captured: 36629
  expected: 100000
  percent: 36.629

```
![plots](https://user-images.githubusercontent.com/16702721/84668202-9cf15800-af23-11ea-95de-72bbf00fbe97.jpg)

adamdbrw commented:
@emersonknapp Just a kind reminder that this is ready for a review

emersonknapp commented:
Thanks for the reminder, we'll start taking a look!

emersonknapp commented on rosbag2_performance/README.md:
```diff
@@ -0,0 +1,44 @@
+- **writer_benchmarking**: the primary package to test transport-less performance of the rosbag2 writer and storage. Enables parametrized batch execution of benchmarks and tries to replicate the flow to capture message loss in queues. This package is the primary one to use when storage optimizations are implemented.
+- **rosbag2_benchmarking**: a python package using launch file and scripting to launch performance workers and a suite of monitor tools (recording cpu, memory and i/o use). Benchmarks the entire pipeline from publish to writing on disk (including the transport layer).   
+- **rosbag2_performance workers**: publishers of random images, point clouds and byte messages, including scaling up and parametrization. Workers random image and pc2 generation part was based on Martin Idel [code](https://github.com/Martin-Idel/rosbag2/tree/performance_testing).
+see raport with `dummy_raport.sh`.
```
When you write "raport", do you mean "report", throughout this code?

adamdbrw replied:
Yes, indeed, a spelling mistake to correct there

Karsten1987 commented on rosbag2_performance/rosbag2_benchmarking/plotting/data_examples/example-data-10Hz-4Mb-0cache.txt:
```diff
@@ -0,0 +1,33 @@
+[INFO] [1591705625.572989661] [rosbag2_transport]: Subscribed to topic '/image'
```
what's the purpose of these cache files? Also I think "[Callback]" is actually not part of the output.

pijaro replied:
These cache files are parts of our preliminary study. Actually the good idea is to remove them as there are no really longer a part of these packages.

Karsten1987 commented on rosbag2_performance/rosbag2_benchmarking/rosbag2_benchmarking/dummy_publishers.py:
```diff
@@ -0,0 +1,79 @@
+# Copyright 2020, Robotec.ai sp. z o.o.
```
why do I need this? What's the difference to `ros2 topic pub <topic> <msg_type>`?

pijaro replied:
Launching `ros2 topic pub` looked pretty slow, what is more, launching 1000 instances was way slower :smiley:

Karsten1987 commented on rosbag2_performance/rosbag2_benchmarking/rosbag2_benchmarking/raport_gen.py:
```diff
@@ -0,0 +1,434 @@
+# Copyright 2020, Robotec.ai sp. z o.o.
```
should that file be renamed to `report`?

pijaro replied:
Yes, it should. Will fix it, thanks :smiley:

Karsten1987 commented on rosbag2_performance/rosbag2_benchmarking/rosbag2_benchmarking/raport_gen.py:
```diff
@@ -0,0 +1,434 @@
+        self.logger.info(str(self.result))
+
+
+def main():
+    """Ros2 once-spin run."""
+    rclpy.init()
+    node = ReportGen()
+    while rclpy.ok():
+        rclpy.spin_once(node, timeout_sec=0.1)
+        rclpy.shutdown()
+    node.destroy_node()
```
why do we need a `rclpy` node for generating a report?

pijaro replied:
You are correct, we should not need `rclpy` for that. It was used to keep the consistency in using running arguments (ie `--ros-args -p foo:=bar`). Will be fixed.

Karsten1987 commented on rosbag2_performance/rosbag2_benchmarking/rosbag2_benchmarking/system_monitor.py:
```diff
@@ -0,0 +1,148 @@
+
+    def __launch_iotop(self):
+        self.dstat_proc = subprocess.Popen(
+            [
+                'pkexec',
+                'iotop',
+                '-b',
+                '-o',
+                '-d',
+                str(1/self.__frequency),
+                '--kilobytes'
+            ],
```
so this wouldn't run on windows. Not that it's super important, but I feel we should have a sanity check here that the exec is actually available.

pijaro replied:
Good point. I will add OS check here.

Karsten1987 commented on rosbag2_performance/rosbag2_benchmarking/scripts/install_pkexec_iotop.sh:
```diff
@@ -0,0 +1,15 @@
+echo ""
+echo "--------------"
+
+sudo cp $SCRIPTPATH/iotop.policy /usr/share/polkit-1/actions/iotop.policy
```
👎  I don't appreciate that we modify users configs here.

pijaro replied:
Yes, indeed we included an `iotop` custom policy here and modifying anything in `usr/share` is controversial at best. However, using this policy is not necessary. It is just a noticeable quality of live improvement for large test bundles. User can still enter password before each test and everything will be fine.

Sudo is required because of the system utilization tool we are using (`iotop`) and this is reasonable because we are literally recording system stats at the time.

We don’t really appreciate that as well so we will look for an different tool which will not require root privileges. Alternatively we can think about moving out the `iotop` instance from single benchmark to something that will run more globally and require only one sudo prompt. I will look into that!

adamdbrw commented:
Thank you for the review! A general comment:
There are two separate packages for benchmarking: full stack, including transport - package rosbag2_benchmarking, and without transport - package writer_benchmarking. These roles are also described in the top-level README file. Both approaches are intended. These packages are quite different and separate from each other - and that would be my first idea to break the PR into pieces. Would you like us to break the PR this way?

@pijaro will handle the review remarks of rosbag2_benchmarking (his package and he is better at Python ;) ), while I handle writer_benchmarking part

pijaro commented on rosbag2_performance/rosbag2_benchmarking/bundles/voyager.sh:
```diff
@@ -0,0 +1,65 @@
+SCRIPT=$(readlink -f "$0")
+SCRIPTPATH=$(dirname "$SCRIPT")
+
+# REPORT_DIR=~/.ros/reports
+
```
To be fixed!

pijaro commented on rosbag2_performance/rosbag2_benchmarking/bundles/voyager.sh:
```diff
@@ -0,0 +1,65 @@
+#!/usr/bin/env bash
```
Definitely a good idea, this would be a way better approach.

pijaro commented on rosbag2_performance/rosbag2_benchmarking/bundles/voyager/1000k_1.yaml:
```diff
@@ -0,0 +1,19 @@
+        delay: 1000
+        max_count: 1000
+        size: 1000000
+        frequency: 100
```
To be fixed!

pijaro commented on rosbag2_performance/rosbag2_benchmarking/launch/benchmarking.launch.py:
```diff
@@ -0,0 +1,419 @@
+import yaml
+
+VERBOSE = True
+logger = launch.logging.get_logger('BENCHMARK')
```
You are right, will be changed!

pijaro commented on rosbag2_performance/rosbag2_benchmarking/launch/benchmarking.launch.py:
```diff
@@ -0,0 +1,419 @@
+
+
+def get_dummy_publisher(topics, types):
+    """Instantiate dummy publisher action for warming up topics."""
```
Warming up topics basically means to let rosbag2 know about every topic we are going to use before we start any worker or monitoring tool. The reasoning behind this is as follows: it makes the synchronization process much cleaner – we are not relying on different workers to pop up the discover on rosbag2, we just want them to produce data as soon as rosbag2 is ready; discovering large number of topics (thousands) at the same time usually leads to some topics not being discovered by rosbag2 – that's why “warming topics” tool is doing that in a more staggered fashion; we tried to be focused as much on utilizing the writing capabilities of rosbag2, not the discovery issues. 

Now I see that this mechanic may be too strange and unnecessary after all. I'm thinking about an alternative checks for topics existence and probably will propose some different and lighter approach.

pijaro commented on rosbag2_performance/rosbag2_benchmarking/launch/benchmarking.launch.py:
```diff
@@ -0,0 +1,419 @@
+                reason='Benchmark done'
+                ))
+
+    # Run dummy publishers to warm up topics when rosbag is ready and
+    # listening for new topics
+    def bag_warm_check(event):
+        target_str = 'Listening for topics...'
+        if target_str in event.text.decode():
+            logger.info('Launching dummy pub')
+            return get_dummy_publisher(worker_topics, worker_types)
+
+    # Run workers when rosbag is warmed up with required topics
+    def bag_run_check(event):
+        target_str = \
+            'All requested topics are subscribed. Stopping discovery...'
+        if target_str in event.text.decode():
+            logger.info('Launching workers')
+            return workers_to_run
```
Not quite, please see @adamdbrw response.

pijaro commented on rosbag2_performance/rosbag2_benchmarking/launch/benchmarking.launch.py:
```diff
@@ -0,0 +1,419 @@
+# Copyright 2020, Robotec.ai sp. z o.o.
```
Yes, this uses transport. Purpose of this launchfile is to run a complete benchmark which includes rosbag2, workers and monitoring tool in a correct sequence. I will add some more documentation to this.

pijaro commented:
Thanks for review, we really appreciate that! We will put some fixes soon and then split this PR into pieces.

Karsten1987 commented:
> These packages are quite different and separate from each other - and that would be my first idea to break the PR into pieces. Would you like us to break the PR this way?

Yes, this would make a lot of sense to me. First, a ROS agnostic benchmark package which tests the storage backends: Ideally, this should not only test sqlite3, but should be possible to test any other backend as well.

Why are these packages different though? I was wondering why browsing through the code, can the package testing transport not rely on the ROS agnostic package? That way, changing things in one more package also clearly affects the other one and doesn't have to be fixed twice.

Generally, it makes the review a bit easier if all files which are no longer needed are removed from this PR.

Karsten1987 commented:
@adamdbrw @pijaro do you have any update on this PR? I have a few ideas on how to address the findings in the performance reports, but they are somehow blocked by this PR. I think we have to have the benchmarking tool in place before  reviewing further performance PRs in order to verify that these are actual improvements.

adamdbrw commented:
@Karsten1987 - I have trimmed down this PR to the single package to streamline our process. The second benchmarking package (which includes transport) will be added separately after this PR review is over.

I have updated this PR description - there is also a list of proposed enhancements, which are known shortcomings of the performance package. Please let me know what is required for this package to be useful to you right away, and what you think can be raised as an enhancement issue. I will try to do my best for this to be merged!

As for this question:
> Why are these packages different though? I was wondering why browsing through the code, can the package testing transport not rely on the ROS agnostic package? That way, changing things in one more package also clearly affects the other one and doesn't have to be fixed twice.

We can wait for the second PR to handle this. In my opinion they are good as independent of each other since the other one does black-box testing - and the common code is already separated out - it is the rosbag2 code.

adamdbrw commented on ros2bag/pytest.ini:
```diff
@@ -0,0 +1,2 @@
+[pytest]
```
Removed. Good catch. Also, have no idea how it got there

Karsten1987 commented on rosbag2_performance/writer_benchmarking/include/message_queue.hpp:
```diff
@@ -0,0 +1,96 @@
+      std::cerr << "X";
+      return;
+    }
+    // std::cerr << "+";
```
```suggestion
```

Karsten1987 commented on rosbag2_performance/writer_benchmarking/include/message_queue.hpp:
```diff
@@ -0,0 +1,96 @@
+  {
+    std::lock_guard<std::mutex> lock(mMutex);
+    if (mQueue.size() > mMaxSize) {  // We skip the element and consider it "lost"
+      mUnsuccessfulInsertCount++;
```
```suggestion
      ++mUnsuccessfulInsertCount;
```

Karsten1987 commented on rosbag2_performance/writer_benchmarking/include/message_queue.hpp:
```diff
@@ -0,0 +1,96 @@
+      throw std::out_of_range("Queue is empty, cannot pop. Check if empty first");
+    }
+    T elem = mQueue.front();
+    // std::cerr << "-";
```
```suggestion
```

Karsten1987 commented on rosbag2_performance/writer_benchmarking/include/message_queue.hpp:
```diff
@@ -0,0 +1,96 @@
+    if (mQueue.empty()) {
+      throw std::out_of_range("Queue is empty, cannot pop. Check if empty first");
+    }
+    T elem = mQueue.front();
```
```suggestion
    auto & elem = mQueue.front();
```

adamdbrw replied:
The of advantage of auto & here isn't clear to me. Since T is a shared_ptr in our case, the difference is neglible (but what you propose wouldn't work with the value type). Could you clarify?

Karsten1987 commented on rosbag2_performance/writer_benchmarking/include/message_queue.hpp:
```diff
@@ -0,0 +1,96 @@
+
+  unsigned int getMissedElementsCount() const
+  {
+    std::lock_guard<std::mutex> lock(mMutex);
```
I think we wouldn't need a lock here if `mUnsuccessfulInsertCount` would be a atomic

Karsten1987 commented on rosbag2_performance/writer_benchmarking/include/writer_benchmark.hpp:
```diff
@@ -0,0 +1,49 @@
+  std::string mDbFolder;
+  std::shared_ptr<rosbag2_cpp::writers::SequentialWriter> mWriter;
+  std::vector<std::thread> mProducerThreads;
+  std::vector<std::shared_ptr<ByteProducer>> mProducers;
+  std::vector<std::shared_ptr<ByteMessageQueue>> mQueues;
```
why are these `shared_ptr`?

adamdbrw replied:
Queues are shared between producers and the writer_benchmark (consumer). They need to be shared_ptrs.
As for ByteProducers, indeed they are owned by writer_benchmark.hpp. Either by value or unique_ptr is better - I will change that

Karsten1987 replied:
+1 for the shared_ptr explanation. As for the unique_ptr, why heap allocation?

adamdbrw replied:
Less prone to errors with element access / copy. Otherwise better by value and accessing by reference. I am fine with either, performance effect is negligible.

Karsten1987 commented on rosbag2_performance/writer_benchmarking/src/main.cpp:
```diff
@@ -0,0 +1,28 @@
+int main(int argc, char * argv[])
+{
+  rclcpp::init(argc, argv);
+  rclcpp::Context ctx;
```
```suggestion
```

Karsten1987 replied:
I would suspect you'd need a `rclcpp::spin(bench)` to actually make the node react to the parameters.

adamdbrw replied:
The node apparently reacts to the parameters given by command line without the need to spin

Karsten1987 commented on rosbag2_performance/writer_benchmarking/src/writer_benchmark.cpp:
```diff
@@ -0,0 +1,189 @@
+WriterBenchmark::WriterBenchmark()
+  : rclcpp::Node("writer_benchmark")
+{
+  this->declare_parameter("frequency");
```
You can directly set the default values: 
```suggestion
  this->declare_parameter("frequency", 100);
```

Karsten1987 replied:
same below for all `declare_parameter` calls.

Karsten1987 commented on rosbag2_performance/writer_benchmarking/src/writer_benchmark.cpp:
```diff
@@ -0,0 +1,189 @@
+  this->declare_parameter("max_cache_size");
+  this->declare_parameter("db_folder");
+
+  auto parameters_client = std::make_shared<rclcpp::SyncParametersClient>(this);
+  while (!parameters_client->wait_for_service(1s))
+  {
+      if (!rclcpp::ok())
+      {
+          RCLCPP_ERROR(this->get_logger(), "Interrupted while waiting for the service. Exiting.");
+          rclcpp::shutdown();
+      }
+      RCLCPP_INFO(this->get_logger(), "service not available, waiting again...");
+  }
+  mConfig.frequency = parameters_client->get_parameter("frequency", 100);
```
you don't need a dedicated parameter client for this.
https://index.ros.org/doc/ros2/Tutorials/Using-Parameters-In-A-Class-CPP/#examine-the-code

Karsten1987 replied:
in fact, I assume we wouldn't need to react dynamically to parameter changes in the first place.

adamdbrw replied:
Correct - I replaced them with this->get_parameter(..) calls

Karsten1987 commented on rosbag2_performance/writer_benchmarking/src/writer_benchmark.cpp:
```diff
@@ -0,0 +1,189 @@
+    for (size_t i = 0; i < mQueues.size(); ++i)
+    {   // TODO(adamdbrw). Performance can be improved. Use conditional variables
+      if (mQueues.at(i)->isComplete() && mQueues.at(i)->isEmpty())
+        completeCount++;
```
```suggestion
        ++completeCount;
```

adamdbrw commented on rosbag2_performance/writer_benchmarking/include/byte_producer.hpp:
```diff
@@ -0,0 +1,80 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef BYTE_PRODUCER_HPP_
```
Actually, now it should be ROSBAG2_PERFORMAMCE_WRITER_BENCHMARKING__BYTE_PRODUCER_HPP_ i believe (with package name changed according to your suggestion) - i will apply this

adamdbrw replied:
Linter does complain - it demands compliance with class name it seems. I have changed it according to your suggestion nevertheless. These includes aren't exported anywhere, so I guess it could stay either way.

adamdbrw commented on rosbag2_performance/writer_benchmarking/include/byte_producer.hpp:
```diff
@@ -0,0 +1,80 @@
+private:
+  std::vector<uint8_t> randomByteArrayData(size_t size)
+  {
+    std::vector<uint8_t> byte(size, 0);
```
what is the reason behind this one? I understand how statics in function work. This doesn't extend the lifetime of the message since the mMessage actually holds the bytes (a copy is made). I am not sure if this would actually mess up RVO in this case (correct me if I am wrong).

The effect is very minor in either case.

Karsten1987 replied:
I had [Meyer's singleton](http://laristra.github.io/flecsi/src/developer-guide/patterns/meyers_singleton.html) in mind here, but as you said it's a minor. RVO is just fine.

adamdbrw commented on rosbag2_performance/writer_benchmarking/include/byte_producer.hpp:
```diff
@@ -0,0 +1,80 @@
+  void run()
+  {
+    for (unsigned int i = 0; i < mConfiguration.max_count; ++i) {
+      if (!rclcpp::ok()) {
+        break;
+      }
```
Main reason is to respond nicely to Ctrl-C without the need to write own signal handler. I believe it is worth an include - however, replacing rclcpp.hpp with include of utilities.hpp seems prudent.

Karsten1987 replied:
I was thinking about removing `rclcpp` all together as a dependency. But as you're using a node in the code below, there's no way around it anyway.

adamdbrw commented on rosbag2_performance/writer_benchmarking/include/byte_producer.hpp:
```diff
@@ -0,0 +1,80 @@
+#include <thread>
+#include <vector>
+#include "message_queue.hpp"
+#include "rclcpp/rclcpp.hpp"
```
See below - I think utilities.hpp is enough (we need ok())

adamdbrw commented on rosbag2_performance/writer_benchmarking/include/byte_producer.hpp:
```diff
@@ -0,0 +1,80 @@
+
+  // for simplification, this pointer will be reused
+  std::shared_ptr<std_msgs::msg::ByteMultiArray> mMessage;
+  std::mutex mMutex;
```
indeed! Removed

adamdbrw commented on rosbag2_performance/writer_benchmarking/include/message_queue.hpp:
```diff
@@ -0,0 +1,96 @@
+    std::lock_guard<std::mutex> lock(mMutex);
+    if (mQueue.size() > mMaxSize) {  // We skip the element and consider it "lost"
+      mUnsuccessfulInsertCount++;
+      std::cerr << "X";
```
Is this primarily for later testing with crashes? There is a performance cost.

Karsten1987 replied:
I was more concerned about thread-safety and having the output in a timely manner. I guess, a real threadsafe print would be `fprintf`

adamdbrw commented on rosbag2_performance/writer_benchmarking/include/byte_producer.hpp:
```diff
@@ -0,0 +1,80 @@
+
+  void generateRandomMessage()
+  {   // Reuses the same random message
+    mMessage = std::make_shared<std_msgs::msg::ByteMultiArray>();
```
As much as I agree with the notion of reducing dependencies, I believe UnboundedSequences.msg is unsuitable for benchmarking. This is simply due to the size of the structure and resulting non-intuitive comparison between performance with message sizes of 1000, 10000 and so on. 

```
test_msgs::msg::UnboundedSequences unbSeq;
std::cerr << "size: " << sizeof(unbSeq) << std::endl;
```
size: 784

IMHO this is too much overhead for reasonable benchmarking of smaller messages.

adamdbrw commented:
> As for style, I know it's not the most important thing here for this package, but consider adhering to the ROS2 style guide: https://index.ros.org/doc/ros2/Contributing/Code-Style-Language-Versions/#id1
> Functions are `snake_case` and private member functions are suffixed with an underscore instead of `mVar`. You might also activate the linters to see that cppcheck and stuff passes.

I ran both ament_cppcheck and ament_cpplint (previously and now again)
The only complain is about the header guards (i.e. BYTE_PRODUCER_HPP_ is preferred by the linter) - which I ignore according to your suggestion.
It doesn't complain about mVar and camelCase for function names. Should I still change them to _var and snake_case?

Karsten1987 commented on rosbag2_performance/rosbag2_performance_writer_benchmarking/include/message_queue.hpp:
```diff
@@ -0,0 +1,94 @@
+  MessageQueue(int maxSize, std::string topicName)
+  : mMaxSize(maxSize), mTopicName(topicName), mUnsuccessfulInsertCount(0) {}
+
+  void push(T elem)
```
if we deal with shared_pointers exclusively, then I would template it accordingly. Either take a `std::shared_ptr<T> elem` as in put or take `T` as `const T & elem` if more generic.

adamdbrw replied:
agreed!

Karsten1987 commented on rosbag2_performance/rosbag2_performance_writer_benchmarking/include/message_queue.hpp:
```diff
@@ -0,0 +1,94 @@
+    if (mQueue.empty()) {
+      throw std::out_of_range("Queue is empty, cannot pop. Check if empty first");
+    }
+    T elem = mQueue.front();
```
same comment as above. If `T` is always meant to be a `shared_ptr<T>` then we should template it accordingly.

Karsten1987 commented on rosbag2_performance/rosbag2_performance_writer_benchmarking/src/writer_benchmark.cpp:
```diff
@@ -0,0 +1,181 @@
+  RCLCPP_INFO(get_logger(), "\nWriterBenchmark terminating");
+  RCLCPP_INFO_STREAM(get_logger(), "Total missed messages: " << totalMissedMessages);
+  RCLCPP_INFO_STREAM(get_logger(), "Percentage of all message that was successfully recorded: "
+    << 100.0 - (float)totalMissedMessages*100.0/(mConfig.max_count*mProducers.size()));
```
```suggestion
    << 100.0 - (float)totalMissedMessages * 100.0 / (mConfig.max_count * mProducers.size()));
```

Karsten1987 commented:
> It doesn't complain about mVar and camelCase for function names. Would you still prefer me to change them to _var and snake_case (no problem)?
It won't complain in this case. It' solely the convention within the ROS2 code base. So it's preferred, but not enforced. Your call ;-)

> As for storing the values in files, the current script (benchmark.sh) outputs a bunch of .log files, which can be processed automatically, although with some grepping.
I've just realized this. I would be very much in favor if we could change this behavior to output the log files directly within the benchmarking code. Writing these data to file within C++ should really be straight forward. Grepping the stderr seems unnecessarily complicated for this and error prone if case other prints are interleaved at some point.

adamdbrw commented:
Alright, time for a second round! We are getting to it :+1: 
Could you look also at my responses to your earlier suggestions? Not everything was merged and perhaps some still should.

adamdbrw commented:
@Karsten1987 take a look at the latest changes:

- applied the remaining two review comments
- renamed member variables and functions to _variable and _snake_case
- implemented logging of results to common `csv` file which is straightforward to plot

With the last change, the results of a benchmark now also reside in a <data> subfolder and there is a summary `csv` file there. I kept the log files as they can serve to diagnose what happened during execution of each benchmark.

Attached: an example file from a run of 10Mb/s benchmarks (changed to `txt` extension because github complains about `csv`)
[results.txt](https://github.com/ros2/rosbag2/files/5015558/results.txt)
:

Karsten1987 commented on rosbag2_performance/rosbag2_performance_writer_benchmarking/README.md:
```diff
@@ -0,0 +1,53 @@
+
+**Using fio**
+
+For more sophisticated & accurate benchmarks, see the `fio` command. An example for big data blocks is: `fio --name TEST --eta-newline-5s --filename=fio-tempfile.dat --rw=write --size=500m --io_size=10g --blocksize=1024k --ioengine=libaio --fsync=10000 --iodepth=32 --direct=1 --numjobs=1 --runtime=60 --group_reporting`.
```
```suggestion
For more sophisticated & accurate benchmarks, see the `fio` command. An example for big data blocks is: `fio --name TEST --eta-newline=5s --filename=fio-tempfile.dat --rw=write --size=500m --io_size=10g --blocksize=1024k --ioengine=libaio --fsync=10000 --iodepth=32 --direct=1 --numjobs=1 --runtime=60 --group_reporting`.
```

adamdbrw commented:
@Karsten1987 I merged the pull request and the README fix. I also added spinning to the main function - it is in a separate thread since the benchmark has its own control loop: `ros2 param list` works correctly now and `colcon test` passes as well.

Is it ready to be merged as far as you are concerned? I understand the custom script is going to be replaced with a neat launch file solution later on, but this is not blocking the merge.

Karsten1987 commented:
how did you test that the parameters are working? I actually still run into the same problem, that while running the batch script, I can't call `ros2 param list` and get the appropriate parameters returned.

adamdbrw commented:
Just running the benchmark.sh script and ros2 param list in the other window...
![image](https://user-images.githubusercontent.com/16702721/89339452-7c7d8880-d69e-11ea-80a2-028985b313b0.png)

Karsten1987 commented:
hmm. I must be doing something wrong then. Also, could you rebase your branch on top of master please?

adamdbrw commented:
Uh.. that went wrong. I will resolve this

Karsten1987 commented:
we'll squash this PR at the end, so if rebasing is too complicated, just merge master into your branch. That should work as well.

adamdbrw commented:
The tree should look good now, though through my error this PR history is difficult to read

Karsten1987 commented:
thanks for rebasing. We have to iterate once more over the bash script. The output directory is being created by rosbag2_cpp itself on master, where the script creates these. The script will now crash.

Second, I would appreciate if we could remap the node name (currently hardcoded to `benchmark_writer`) to reflect the current running instance to avoid any node conflicts.

adamdbrw commented:
Done - please take a look if this is what you wanted

Karsten1987 commented on rosbag2_performance/rosbag2_performance_writer_benchmarking/src/writer_benchmark.cpp:
```diff
@@ -0,0 +1,216 @@
+
+void WriterBenchmark::write_results(const unsigned int & total_missed) const
+{
+  unsigned int total_messages_sent = config_.max_count * producers_.size();
```
```suggestion
  auto total_messages_sent = config_.max_count * producers_.size();
```

Karsten1987 commented:
Windows once more: [![Build Status](https://ci.ros2.org/buildStatus/icon?job=ci_windows&build=11724)](https://ci.ros2.org/job/ci_windows/11724/)
